src/extensions/client/health.rs

```diff
@@ -0,0 +1,151 @@
+    pub fn new(url: String, config: HealthCheckConfig) -> Self {
+        Self {
+            url,
+            config,
+            score: AtomicU32::new(100),
+            unhealthy: tokio::sync::Notify::new(),
+        }
+    }
+
+    pub fn score(&self) -> u32 {
+        self.score.load(Ordering::Relaxed)
+    }
+
+    pub fn update(&self, event: Event) {
+        let current_score = self.score.load(Ordering::Relaxed);
+        let new_score = u32::min(
+            match event {
+                Event::ResponseOk => current_score.saturating_add(5),
+                Event::SlowResponse => current_score.saturating_sub(5),
+                Event::RequestTimeout | Event::ConnectionFailed | Event::StaleChain => 0,
+                Event::ConnectionSuccessful => 100,
+            },
+            MAX_SCORE,
+        );
+        self.score.store(new_score, Ordering::Relaxed);
+        // Notify waiters if the score has dropped below the threshold
+        if current_score >= THRESHOLD && new_score < THRESHOLD {
+            self.unhealthy.notify_waiters();
+        }
+    }
+
+    pub fn on_error(&self, err: &jsonrpsee::core::Error) {
+        match err {
+            jsonrpsee::core::Error::RequestTimeout => {
+                self.update(Event::RequestTimeout);
+            }
+            jsonrpsee::core::Error::Transport(_)
+            | jsonrpsee::core::Error::RestartNeeded(_)
+            | jsonrpsee::core::Error::MaxSlotsExceeded => {
+                self.update(Event::ConnectionFailed);
+            }
+            _ => {}
+        };
+    }
+}
+
+impl Health {
+    pub fn monitor(
+        health: Arc<Health>,
+        client_rx_: tokio::sync::watch::Receiver<Option<Arc<Client>>>,
+        on_client_ready: Arc<tokio::sync::Notify>,
+    ) -> tokio::task::JoinHandle<()> {
+        tokio::spawn(async move {
+            // Wait for the client to be ready before starting the health check
+            on_client_ready.notified().await;
+
+            let method_name = health.config.health_method.as_str();
+            let interval = Duration::from_secs(health.config.interval_sec);
+            let healthy_response_time = Duration::from_millis(health.config.healthy_response_time_ms);
+
+            let client = match client_rx_.borrow().clone() {
+                Some(client) => client,
+                None => return,
+            };
+
+            loop {
+                // Wait for the next interval
+                tokio::time::sleep(interval).await;
+
+                let request_start = std::time::Instant::now();
+                match client
+                    .request::<serde_json::Value, Vec<serde_json::Value>>(method_name, vec![])
+                    .await
+                {
+                    Ok(response) => {
+                        let duration = request_start.elapsed();
+
+                        // Check known health responses
+                        match method_name {
+                            "system_health" => {
```

Comment:
this is fine but it will be better if we can config this instead of hardcode them here